refactored to combine collector logic into samples to better utilize it

kkim-sf opened the pull request:
- Removed the existing collector class which was hidden away not utilizing the bulk ability (samples attr was never getting populated with more than one array item)
- Added upsert, add_to_bulk, upsert_bulk, and upsert_custom_body methods to the Samples class to provide some flexbility
- upsert_bulk logic from the original Collector class's 'submit' method updated to return the actual response as oppsed to always returning true
- Made the "value" parameter an optional parameter as the API document also specifies it as an optional field.

Commit:
refactored to combine collector logic into samples to better utilize it

sheax0r commented on lib/refocus/samples.rb:
```diff
@@ -5,20 +5,35 @@ module Refocus
+        samples.clear
+        result
+      end
+    end
```
@kkim-sf First of all: Thanks for helping out and welcome! :)

I'd like to elaborate a bit on the pattern I'm promoting here. It looks like in order to use what you've done effectively, you'll need to do something like this:

```ruby
  refocus.samples.add_to_bulk(name:"mysample", aspect: "myaspect", value: 1)
  refocus.samples.add_to_bulk(name:"mysample", aspect: "myaspect", value: 2)
  refocus.samples.upsert_bulk
```

The problem I have with this is that this is that the collector already enables this sort of semantics - ie, "Add samples one at a time, then submit all at once", like this:

```ruby
   collector = refocus.samples.collector
   collector.add(name: "mysample", aspect: "myaspect")
  collector.submit
```

I'd rather not have the client itself be responsible for maintaining any state other than knowledge of the endpoint it is talking to. The collector should be used to store state like that, which is what it does.

I like the addition of the new parameters - I just think the should be added to the existing `sample.submit` method, and passed through to `Collector.submit`. Submission of a single sample could also be done using its own method in the Samples class instead of passing the values through to a one-off collector object, but I don't think it really matters much for efficiency (ultimately, it would hit the same API). I implemented it this way to avoid rewriting the same code to submit samples twice.

Btw, feel free to reach out on gchat or slack if you want more synchronous communication! I'm in Australia though, so we might have trouble with timezones.

Also - value is optional? Good to know, but it strikes me as strange. What does it even mean to have a sample without a value? 

If you prefer `upsert` instead of `submit` for the method name, I propose maintaining compatibility by having both, perhaps with `alias_method`.

kkim-sf committed:
reverted Collection. added readme for Samples

kkim-sf committed:
updated readme desc and removed samples attr_reader from Samples class

kkim-sf committed:
added back collector dependency require

kkim-sf committed:
updated readme example hash

kkim-sf commented:
Hi @sheax0r ,

Thanks for the feedback. I'm in NY so it's hard to catch you at the right time on hangouts due to the time difference. 

Per your comments, I've reverted back the Collector class removal. I think I saw it as a bit odd because it was the only endpoint that had its own Collector class whereas the other ones didn't. In addition, my dumb self didn't realize the usage until I saw the implementation again this morning with your example of the collector use -- to my defense, I did not have the documentation on the ReadMe, and it was late at night! ha, I know excuses excuses.

In terms of the method name, I've added the submit aliases so that it's kept backwards compatible. My reasoning for renaming to 'upsert' is an attempt to keep things somewhat consistent -- the other client classes have methods such as create, update, delete, and etc -- and 'submit' didn't seem to be straight forward in alluding to the fact that it is doing 'create or update'.

In terms of the value being the optional param, I think it's actually correct because you're not always creating a new value via this call. You could be updating any one of or combination of the params of an existing sample: value, messageCode, messageBody, and relatedLinks -- and you want to do so without having to update the value every time.

Also, I understand your reasoning that for single items, you could still use the upsert/bulk endpoint, but I much prefer actually hitting the upsert endpoint for single items because the response you get back is the updated sample object whereas the bulk call just returns {status: "ok"}.

I've made some updates to the PR, please take a look!

sheax0r reviewed (approved):
@kkim-sf Hello! No worries about failing to apprehend some of my reasoning in the original version. The documentation and tests for this library are sorely lacking, so thank you for taking the time to wade through it, figure out how it works, and make changes. I really appreciate it. 

This project definitely needs better documentation and tests. I wrote it originally as a way to help me understand refocus at all, and to help me do some stuff I needed in order to get my "real" project done - so it's very bare-bones, and is lacking productionization.

I understand your desire to have the method names reflect the underlying API calls (ie: "upsert_bulk"), and that totally makes sense and is good. Sometimes I try to put my own abstractions and names for things on top of APIs, and it can make things messy when the API turns out to not have the semantics that I imagined it to have. 

I'll release this shortly after merging. This project also doesn't really have a formal release process yet, but so far I feel like it's used little enough that it doesn't need one. One of these days it'll need some semantic versioning though if people actually start to use it.

Thanks so much for contributing!

sheax0r commented:
@kkim-sf released in `v0.1.6`!
